### DISCUSSION TITLE
Wavetable support for OneCycleWaveform

### DISCUSSION THREAD

**Discussion by ryukau** [ryukau/UhhyouWebSynthesizers#2] [Ideas]
This is a requested feature from a user.

What to do:
- Join several renderings to one file.
- Randomize each rendering.

There's a synth called Wavetable in Live. It can take a wave file with length of `1024 * n` samples. I'm not sure if this is accurate, but it seems like each 1024 sample block is assigned to a wavetable frame.

Preparing test environment is the current blocker. I don't have wavetable synth that can read 2 or more frames of waveforms. I'd like to test the result in some way.

For randomization, maybe add something like `random amount`, and use that value for every randomization. The scaling for each parameter becomes a problem. Because it takes time to tune them all.

**Comment by NeuroNoNeuro** [comment]
Possibly offer an option of:

1.  "Create Wavetable"
2. Select quantity of 2 to 100 OneCycleWaveform renderings
3. Press the "Random" button
4. Output a single audio file with 100 OneCycleWaveforms 'glued' together

If possible, the arrangement of renderings could go from soft (sine wave) to harsh (noise) in order.  Like so:

![Screenshot 2023-10-03 at 3 55 45 AM](https://github.com/ryukau/UhhyouWebSynthesizers/assets/38725630/45873901-9853-4faa-9797-1552cfda6707)

This output would be ideal, but even a randomized wavetable would be exciting, also creating much less work, not having to organize the arrangement of waveform timbre.

**Comment by ryukau** [comment]
Sorting spectra is not so straight forward. But in this case, something like spectral whiteness can be used.

**Comment by ryukau** [comment]
Random multiple table generation is added to OneCycleWaveform. 

- `nTable` is the number of wavetables.
- It will randomize only when `Random Amount` is not 0.

@NeuroNoNeuro Let me know if it works on Wavetable.

**Reply by NeuroNoNeuro** [reply]
Oh WOW!!!  I was just going to build some more wavetables by hand, and now this is here.  Saves me a literal TON of time.  Works great!  I just messed with it a little, and chose to manipulate only:

N-Table:  10
Seed:  0
Amount:  25 

Works great!  Holy cow dude.  Exported the file, dropped it in and BOOM.  Done.  Perfect.

**Reply by ryukau** [reply]
I'm glad it worked! 😄

For the record, sorting is somewhat arbitrary. There's a opportunity for different musical tunings.

**Reply by NeuroNoNeuro** [reply]
One question - how would one go about creating additive-type waveforms - I know it can, because I get some through random generation.  What would be a way to get a rudimentary set of additive tones - from sine wave up to varying sorts of sine-waves with odd pulses?

**Reply by ryukau** [reply]
Could you provide an example output? However, I probably have no answer for that. Maybe `Spectral` and `Filter` sections are related to it.

One problem is that the sorting is flaky, that the result doesn't always sort from sine (less overtones) to pulse/glitch (more overtones). So it's hard to expect the result. Other is the parameter tunings on randomization. For this, I need testing environment but I don't have one. Below is the link to the relevant code.

https://github.com/ryukau/UhhyouWebSynthesizers/blob/86ae178bc4c0daf67c7551f66eccb10c37912516/OneCycleWaveform/renderer.js#L64-L93

**Reply by NeuroNoNeuro** [reply]
Unfortunately I cannot.  I can say that your PADchoir synth can create the waveforms if the user selects ‘additive pad’.

**Comment by RossoRacer** [comment]
I have encountered an issue using One Cycle. It’s to do with looping single cycle waves. After saving a .wav I would use Endless Wav to set its loop to on and then I would be able to use the .wav in Audioterm. However, One Cycle appears to do something different. If you open the .wav in Endless Wav it’s sort of got half way saved as a loop on. What I mean is that the loop on indicator in Endless Wav is half lit, ie neither on or off. You can try and save it as on but Audioterm will not recognise the .wav because the loop is effective off. If you try to open a .wav saved directly from One Cycle in Audioterm again, the same problem. It’s unrecognised as a file. Any suggestions?

**Reply by ryukau** [reply]
I added `Loop Point` parameter to OneCycleWaveform (Render section, mid left). Perhaps turning off Loop Point might work for your applications. If it doesn't work, please provide a working .wav file. It requires to look into the file header.

---

OneCycleWaveform outputs 32-bit float wav. Sampling rate depends on your system default setting, that is read by your web browser and provided to my synthesizers through [`AudioContext.sampleRate`](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/AudioContext). It might be better to check if this wav file format is supported by Endless Wav and Audioterm.

Loop point is set as `smpl` chunk. However, it's only tested on FL because I couldn't find a reliable way to test it. So my code might be incorrect here. Below is the link to the relevant code (common/wave.js, `fileHeader()`).

https://github.com/ryukau/UhhyouWebSynthesizers/blob/540f418fc480f3497cc52ab8d6aec753a73c5f4a/common/wave.js#L296-L411

**Comment by RossoRacer** [comment]
By the way, truly excellent utility. Thanks for taking the time to make it. Likewise the others too!

I did get one .wav to work properly but I’m not entirely sure how at this point. I’ll do some more tests and get back to you when I have consistent results. I’m creating for Waldorf Blofeld, hence using AudioTerm.

**Comment by RossoRacer** [comment]
I have now figured out what the issue was and it wasn’t OneCycle.

I thought it might be changing the frame size down to 128 samples by typing in manually, but that was not it. A combination of an unstable version of Audacity (3.7) and other factors were to blame. After reading about the OneCycle output being 32 bit float I realised I’d have to run each .wav through Audacity first to get it to a 16 bit .wav before AudioTerm would recognise it. Then each single cycle will have to be run through Endless Wav because Audacity won’t let you set loop to on. Anyway, it’s somewhat convoluted but it works! Looking forward now to hours of experimentation.

Thanks again for creating this. It’s really useful.
